pkg/util/push/error/errors.go

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package pushpb
+
+import (
+	"slices"
+
+	"github.com/gogo/status"
+	"google.golang.org/grpc/codes"
+	grpc "google.golang.org/grpc/status"
+)
+
+type ErrorDetail interface {
+	errorDetail()
+}
+
+func (t ErrorType) errorDetail() {}
+
+func (f OptionalFlag) errorDetail() {}
+
+func (p *PushErrorDetails) ContainsOptionalFlag(optionalFlag OptionalFlag) bool {
+	return slices.Contains(p.GetOptionalFlags(), optionalFlag)
+}
+
+// newPushErrorDetails creates a new instance of PushErrorDetails by using
+// the given set of ErrorDetail objects.
+func newPushErrorDetails(errorDetails ...ErrorDetail) *PushErrorDetails {
+	var (
+		errorType     ErrorType
+		optionalFlags []OptionalFlag
+	)
+	if len(errorDetails) > 1 {
+		optionalFlags = make([]OptionalFlag, 0, len(errorDetails)-1)
+	}
+	for _, opt := range errorDetails {
+		eT, ok := opt.(ErrorType)
+		if ok {
+			errorType = eT
+			continue
+		}
+		oF, ok := opt.(OptionalFlag)
+		if ok {
+			optionalFlags = append(optionalFlags, oF)
+		}
+	}
+	return &PushErrorDetails{
+		ErrorType:     errorType,
+		OptionalFlags: optionalFlags,
+	}
+}
+
+type errorWithStatus struct {
+	err    error
+	status *status.Status
+}
+
+func newErrorWithStatus(code codes.Code, originalErr error) errorWithStatus {
+	return errorWithStatus{
+		err: originalErr,
+		status: status.New(
+			code,
+			originalErr.Error(),
+		),
+	}
+}
+
+func (e errorWithStatus) Error() string {
+	return e.status.Message()
+}
+
+func (e errorWithStatus) Unwrap() error {
+	return e.err
+}
+
+func (e errorWithStatus) GRPCStatus() *grpc.Status {
+	se, ok := e.status.Err().(interface{ GRPCStatus() *grpc.Status })
+	if !ok {
+		return nil
+	}
+	return se.GRPCStatus()
+}
+
+// NewPushError gets an error, a gRPC code and a set of ErrorDetail objects,
+// and does the following things:
+//   - if the given error is produced by gogo/status or grpc/status packages,
+//     the error itself with no additional details is returned.
+//   - Otherwise, a new gRPC errorWithStatus error backed up by the given error,
+//     with the given gRPC code and with given set of details is created and
+//     returned.
+func NewPushError(code codes.Code, originalErr error, opts ...ErrorDetail) error {
```

Comment:
I can see `opts` is variadic, but in which case we may have to pass multiple details? What's the use case for that? Also, what's the case for having no error detail at all (currently supported)?

Reply:
>Why do we need to pass the gRPC code too? When we pass our own specific error, I would rather infer the gRPC code from that specific error.

In the current implementation of the distributor the errors created by `Push()`:
- are created by `httpgrpc.Errorf()` and contain an HTTP status code and a message, or
- are generic errors (either created by `errors.New()`, `fmt.Errorf()`, or `context.Canceled`)

The errors returned by ingester's `Push()` have similar features, except for the instance-limit errors, which contain a gRPC code.

The approach I am proposing is that both the distributor (in this PR) and the ingester (in a follow-up) will return errors containing a gRPC code and the underlying error that caused the problem.

Reply:
>I can see opts is variadic, but in which case we may have to pass multiple details? What's the use case for that? 

Depending on the `-distributor.service-overload-status-code-on-rate-limit-enabled` CLI flag, request rate limit errors are reported as a `429` or as a `529`. I tried to put that flag into the push handler, but I didn't manage to do it (it might also be that it is possible, but that I wasn't able to do it), so I added `OptionalFlag`s to the error to help us handle these situations.

> Also, what's the case for having no error detail at all (currently supported)?

context.Canceled